core/src/avm2/class.rs

```diff
@@ -108,6 +110,44 @@ fn do_trait_lookup<'gc>(
     Ok(())
 }
 
+/// Find traits in a list of traits matching a slot ID.
+///
+/// This function also enforces final/override bits on the traits, and will
+/// raise `VerifyError`s as needed.
+///
+/// TODO: This is an O(n^2) algorithm, it sucks.
+fn do_trait_lookup_by_slot<'gc>(
+    id: u32,
+    known_traits: &mut Vec<Trait<'gc>>,
+    all_traits: &[Trait<'gc>],
+) -> Result<(), Error> {
```

Comment:
I'm not sure I follow why we must to return a list of traits for a given slot ID... Every slot ID must be unique for a given object, and they can't be overridden/conflict AFAIK? If you hand edited an ABC to give two traits the same slot ID, you get a VerifyError right off the bat.

The slots allocate out of the memory of the the object, laid out sequentially after another like a C++ object; and the traits objects gives you the mapping from slot_id -> memory offset. There's not a chain of slots shadowing other slots?

Here's an example (hand-edited SWF):
[slot-override-error.zip](https://github.com/ruffle-rs/ruffle/files/6371705/slot-override-error.zip)
`VerifyError: Error #1053: Illegal override of c in Base.`

Reply:
So, the reason why trait lookup here takes a `Vec` is so that we can do verification at trait install time. We do the same thing in the by-name lookup methods.

I can remove the verification here for now, as you pointed out in another comment that the trait override/final bits don't apply to slot numbers. However, we don't really have a good place to enforce the illegal slot override error for now.

Perhaps I should overhaul verification in a future PR...

Reply:
I guess I don't really follow what it means to "install a trait"... it seems like it's designed to lazily put traits into an object, but an object should know all of its traits at once when instantiated. The lazy init flag works on a script level, not at the trait level, AFAIK? It delays the execution of the script until an op tries to grab a definition from the file, but the script is still parsed when the tag is seen, and all of the trait info is available. So verification should be happening when the script is parsed, when the scriptinit runs, and/or when `newclass` is called.

(Again, sorry as I know this is all separate from this PR somewhat; but I'm getting caught on all the AVM2 internals.)

Reply:
So, you are more or less correct; we should be creating all of the object's properties and slots when we allocate the object (or at least, before instance initializers run) rather than when they're first accessed.

The reason why our interpreter lazily instantiates traits is mostly a historical curio; the way I got to the object model we currently have was by starting with an ES3-style object model and adding more AVM2 functionality on top. That's why instances are linked to classes through `__proto__`, for example: I had first rigged up traits as prototype properties, and then switched to having instances get their own traits. But I still used the proto chain to grab the class and handle `derive`; two things that I haven't yet taken out of the prototype and into `Class`. Same thing with lazy init - at the time I needed traits to work, I sort of just assumed that they were created when they were first accessed, which is probably also very wrong.

I should probably have an issue for "all the things that are wrong with our current object model" or something...

Reply:
Okay, thanks for clarifying; making sure my understanding is correct, too.